x/wasm/keeper/keeper.go

```diff
@@ -1076,3 +948,39 @@ func (c BankCoinTransferrer) TransferCoins(ctx sdk.Context, fromAddr sdk.AccAddr
 	}
 	return nil
 }
+
+type msgDispatcher interface {
+	DispatchMessages(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, msgs []wasmvmtypes.CosmosMsg) error
+	DispatchSubmessages(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, msgs []wasmvmtypes.SubMsg) ([]byte, error)
+}
+
+// DefaultWasmVMContractResponseHandler default implementation that first dispatches submessage then normal messages.
+// The Submessage execution may include an success/failure response handling by the contract that can overwrite the
+// original
+type DefaultWasmVMContractResponseHandler struct {
+	md msgDispatcher
+}
+
+func NewDefaultWasmVMContractResponseHandler(md msgDispatcher) *DefaultWasmVMContractResponseHandler {
+	return &DefaultWasmVMContractResponseHandler{md: md}
+}
+
+// Handle processes the data returned by a contract invocation.
+func (h DefaultWasmVMContractResponseHandler) Handle(
+	ctx sdk.Context,
+	contractAddr sdk.AccAddress,
+	ibcPort string,
+	submessages []wasmvmtypes.SubMsg,
+	messages []wasmvmtypes.CosmosMsg,
+	origRspData []byte,
+) ([]byte, error) {
+	result := origRspData
+	switch rsp, err := h.md.DispatchSubmessages(ctx, contractAddr, ibcPort, submessages); {
```

Comment:
After researching `wasmvm`, I came to the conclusion this all works perfectly fine with the existing Go-Rust interface

Reply:
ja, I realized later that it may not be easy to review this as it is displayed as new content. Apologies for this. Tests were not refactored though.

> Note: do we need to differentiate between nil and []byte{} here

If we consider empty `[]bye` a valid response from a contract then we need to be able to overwrite this by the reply. The `nil` result is considered an opt out to overwrite the response.

> After researching wasmvm, I came to the conclusion this all works perfectly fine with the existing Go-Rust interface

👏